test/types.jl

```diff
@@ -236,6 +237,18 @@ end
     @test eltype(broadcast(Gray, rand(Float32,5))) == Gray{Float32}
 end
 
+@testset "transparent gray constructors" begin
+    for x in (1, N0f8(0.5), 0.5f0, 0.5)
+        T = isa(x, Int) ? N0f8 : typeof(x)
+        @test @inferred(AGray(x)) === @inferred(AGray(x, 1)) === AGray{T}(x, 1)
+        @test @inferred(GrayA(x)) === @inferred(GrayA(x, 1)) === GrayA{T}(x, 1)
+        g = Gray(x)
+        @test @inferred(AGray(g)) === @inferred(AGray(g, 1)) === AGray{T}(x, 1)
+        @test @inferred(GrayA(g)) === @inferred(GrayA(g, 1)) === GrayA{T}(x, 1)
+    end
+    @test AGray32(Gray(0.5)) === AGray32(Gray(0.5), 1) === AGray32(N0f8(0.5)) === AGray32(N0f8(0.5), 1) === reinterpret(AGray32, 0xff808080)
+end
```

Comment:
What is the purpose of adding these?

Reply:
Without the change to `src/types.jl` above, `x = N0f8(0.5); AGray(x, 1)` returns an `AGray{Float32}` because `promote(x, 1)` chooses `Float32`. But with an integer argument we know the only valid settings of `alpha` are 0 and 1, so a valid implementation would be to convert it a `Bool` first (drawing the error if that fails) and then call `promote`. But note that `promote(x, true)` chooses `N0f8`, so the final type produced is strongly dependent on internal details of the implementation of the constructor.

To me it seems that a more agnostic approach is to first convert `alpha` to a supported eltype and then do the promotion. Note that changing the signature of those same to methods to `alpha::GrayLike` would circumvent the error you noted in https://github.com/JuliaGraphics/ColorTypes.jl/pull/177#issuecomment-617597990. Though I don't feel strongly that we need to support `Gray` values for the alpha channel, making it consistently throw an error would also be OK.

Reply:
I finally got the purpose. :+1:  (For such a purpose, I prefer to write the types extensionally, though.)

> But with an integer argument we know the only valid settings of `alpha` are 0 and 1

I agree for the most part. However, so far, we have not checked the range of valid values ​​for each "color" component. Actually , we only check the valid range of `Normed` "type", so `AGray{N1f7}(1, 2)` does not throw errors.
Therefore, I prefer the more agnostic approach.

Reply:
> I don't feel strongly that we need to support `Gray` values for the alpha channel,

I agree. However, it is not only problem with the alpha channel.
```julia
julia> RGB{N0f8}(Gray(1), 1, 1)
RGB{N0f8}(1.0,1.0,1.0)

julia> ARGB{N0f8}(Gray(1), 1, 1, 0.5)
ERROR: MethodError: no method matching ARGB{Normed{UInt8,8}}(::Gray{Normed{UInt8,8}}, ::Int64, ::Int64, ::Float64)
```
"Unfortunately", the possible way to make them consistent may be to enhance the support for `GrayLike`, and I support the status quo. :sweat_smile:

Reply:
I do think it's reasonable to construct an RGB from gray channels, as it is the inverse operation of color separation.